clang/lib/Sema/Sema.cpp

```diff
@@ -1670,15 +1677,22 @@ class DeferredDiagnosticsEmitter
       return;
     bool HasWarningOrError = false;
     bool FirstDiag = true;
-    for (PartialDiagnosticAt &PDAt : It->second) {
+    for (Sema::DeviceDeferredDiagnostic &D : It->second) {
       // Respect error limit.
       if (S.Diags.hasFatalErrorOccurred())
         return;
-      const SourceLocation &Loc = PDAt.first;
-      const PartialDiagnostic &PD = PDAt.second;
+      const SourceLocation &Loc = D.getDiag().first;
+      const PartialDiagnostic &PD = D.getDiag().second;
+      Sema::DeviceDiagnosticReason Reason = D.getReason();
       HasWarningOrError |=
           S.getDiagnostics().getDiagnosticLevel(PD.getDiagID(), Loc) >=
           DiagnosticsEngine::Warning;
+
+      // If the diagnostic doesn't apply to this call graph, skip this
+      // diagnostic.
+      if (static_cast<int>(RootReason & Reason) == 0)
```

Comment:
Does this still need a static-cast?  I thought the bitmask macro meant you didn't have to do this anymore?

Reply:
```suggestion
      if (RootReason & Reason == Sema;:DiagnosticReason::None)
```

?

Reply:
Done